remove RX as the primary pipe for gRPC communications

mgravell opened the pull request:
### Issue describing the changes in this PR

resolves #8280; performance locally is a 5x+ speedup in throughput

the main crux of the new flow is described in src/WebJobs.Script.Grpc/Eventing/GrpcEventExtensions.cs, but in short: we establish a per-worker `Channel<InboundGrpcEvent>` and `Channel<OutboundGrpcEvent>` to handle the backlog; instead of publishing to RX, we publish to the writer of the relevant queue. Separately, we have an async worker deque data from the queues, and process accordingly. This is much more direct, avoids a *lot* of RX machinery, and creates isolation between workers.

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Additional PR information

mgravell committed:
remove RX as the primary pipe for gRPC communications

remove RX as the primary pipe for gRPC communications

Commit:
fixup test failures
1: GrpcWorkerChannelTests span up duplicate channels
2: suppress write failures (async void)

mgravell committed:
tweak the worker channel process flow to ensure it is testable; fixup TestFunctionRpcService and GrpcWorkerChannelTests to match

fabiocav commented:
@mgravell apologies for the delay on this. We'll get eyes on this ASAP. Would it be possible to have this rebased to deal with the conflicts? 

Thanks!

mgravell commented:
Will do

On Fri, 29 Apr 2022, 22:21 Fabio Cavalcante, ***@***.***>
wrote:

> @mgravell <https://github.com/mgravell> apologies for the delay on this.
> We'll get eyes on this ASAP. Would it be possible to have this rebased to
> deal with the conflicts?
>
> Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Azure/azure-functions-host/pull/8281#issuecomment-1113745563>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEHMENKREBPYEY25NFYVTVHRHEHANCNFSM5SPTV72Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

pragnagopa commented:
Tagging @safihamid @paulbatum  - fyi 

@fabiocav  / @brettsam  - would be good to get test E2E on private stamp before merging this PR

safihamid commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannelFactory.cs:
```diff
@@ -47,6 +48,7 @@ public IRpcWorkerChannel Create(string scriptRootPath, string runtime, IMetricsL
                 throw new InvalidOperationException($"WorkerCofig for runtime: {runtime} not found");
             }
             string workerId = Guid.NewGuid().ToString();
+            _eventManager.AddGrpcChannels(workerId); // prepare the inbound/outbound dedicated channels
```
@pragnagopa can you confirm this is being called during placeholder mode and shouldn't be impacting cold start?

Also how is this going to impact isolated cold start when there are no placeholders?

liliankasem replied:
@pragnagopa poke for update :)

brettsam replied:
@safihamid -- Can you take this and run it in a private stamp to make sure there's no cold start regressions?

safihamid replied:
ok I don't see any changes in cold start in my private stamp with this. 

this method (_eventManager.AddGrpcChannels(workerId))gets called during placeholder mode.

liliankasem commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -1015,5 +1165,75 @@ private void AddSample<T>(List<T> samples, T sample)
+                {
+                    try
+                    {
+                        throw new TimeoutException();
```
Do we need to throw an new exception just to catch it and handle? Could we just do `_faultHandler(timeout);` in the first try?

mgravell committed:
Merge branch 'dev' into mgravell/remove-rx

# Conflicts:
#	src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs
#	test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs
#	test/WebJobs.Script.Tests/Workers/Rpc/TestFunctionRpcService.cs

mgravell committed:
Merge branch 'dev' into mgravell/remove-rx

mgravell committed:
merge removal of PublishFunctionLoadResponsesEvent work

mgravell commented:
@fabiocav  I *believe* that's merged; it looks like some cheese moved around the topic of "function load collections", which made the tests merge a bit messy, but: should work

brettsam commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
 
+        private void ProcessItem(InboundGrpcEvent msg)
+        {
+            // note this is a QUWI entry-point
```
We should spell out `QueueUserWorkItem` here -- I had to look it up :-)

mgravell replied:
fixing

brettsam commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
+                        {
+                            _workerChannelLogger.LogDebug("[channel] received {0}: {1}", msg.WorkerId, msg.MessageType);
+                        }
+                        ThreadPool.QueueUserWorkItem(_processInbound, msg);
```
Curious: why this over `Task.Run()`?

mgravell replied:
They're both just "run me a thing" abstractions; QUWI works fine, and has fewer overheads (marginally) than Task.Run; either should work

brettsam commented on src/WebJobs.Script.Grpc/Eventing/IGrpcEventManager.cs:
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Remove?

liliankasem commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -99,31 +102,22 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
-            _eventSubscriptions.Add(_inboundWorkerEvents
-                .Where(msg => msg.IsMessageOfType(MsgType.RpcLog) && msg.IsLogOfCategory(RpcLogCategory.System))
-                .Subscribe(SystemLog));
+            if (eventManager.TryGetGrpcChannels(workerId, out var inbound, out var outbound))
```
Should we use `_eventManager` here?

mgravell replied:
no real benefit either way; happy to change, but... we know they're the same value

liliankasem commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -99,31 +102,22 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
+                // note: we don't start the read loop until StartWorkerProcessAsync is called
+            }
 
             _eventSubscriptions.Add(_eventManager.OfType<FileEvent>()
```
Do we also want this to be handled differently? I think it's the only one still using `_eventSubscriptions`

liliankasem commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
+                switch (msg.MessageType)
+                {
+                    case MsgType.RpcLog:
+                        switch (msg.Message.RpcLog.LogCategory)
```
nit: might be cleaner to use the new switch expression: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression

mgravell replied:
switch expressions require ... an expression - i.e. a single computed value for each outcome, that must all evaluate to some single type (perhaps after coercion). In this case, most of the things we actually do here are `void` returns. Forcing it to try to use a switch expression would be unnatural. If you mean a `when` clause, I could perhaps get behind that; as below:

``` c#
                switch (msg.MessageType)
                {
                    case MsgType.RpcLog when msg.Message.RpcLog.LogCategory == RpcLogCategory.System:
                        SystemLog(msg);
                        break;
                    case MsgType.RpcLog:
                        Log(msg);
                        break;
                    case MsgType.WorkerStatusResponse:
                        ReceiveWorkerStatusResponse(msg.Message.RequestId, msg.Message.WorkerStatusResponse);
                        break;
                    case MsgType.InvocationResponse:
                        _ = InvokeResponse(msg.Message.InvocationResponse);
                        break;
                    default:
                        OnNext(msg);
                        break;
                }
```

liliankasem replied:
Yup, that looks great - thanks!

liliankasem commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
+                }
+                for (int i = 0; i < count; i++)
+                {
+                    var newItem = i == count - 1 && timeout != TimeSpan.Zero
```
nit: for readability
```suggestion
                    var newItem = (i == count - 1) && (timeout != TimeSpan.Zero)
```

liliankasem commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -827,9 +937,53 @@ internal void HandleWorkerMetadataRequestError(Exception exc)
+                    }
+                    catch
+                    {
+                        // suppress failure
```
Just wondering why we want to suppress the failure here

liliankasem commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -1002,5 +1158,75 @@ private void AddSample<T>(List<T> samples, T sample)
             }
         }
+
+        private sealed class PendingItem
```
Is this essentially a pending grpc message / event we want to handle? I think a more descriptive name could be helpful. If we change the name here it might also be good to make sure other methods such as `ProcessItem` match i.e. ProcessGrpcEvent

Also wondering if we should move this into it's own file

liliankasem commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -217,6 +329,11 @@ internal void SendWorkerInitRequest(GrpcEvent startEvent)
+            // advertise that we support multiple streams, and hint at a number; with this flag, we allow
+            // clients to connect multiple back-hauls *with the same workerid*, and rely on the internal
+            // plumbing to make sure we don't process everything N times
+            initRequest.Capabilities.Add(RpcWorkerConstants.MultiStream, "10"); // TODO: make this configurable
```
Can we make the number ("10") a readonly constant for now?

Is the TODO for this PR or future work? If future work, can we create a github issue for this and mention it here?

liliankasem commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -827,9 +937,53 @@ internal void HandleWorkerMetadataRequestError(Exception exc)
+                }
+                catch
+                {
+                    // nowehere to log
```
```suggestion
                    // no where to log
```

liliankasem commented on src/WebJobs.Script.Grpc/Eventing/IGrpcEventManager.cs:
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Wondering why this is commented out and if we have plans to use this in this PR?

mgravell replied:
sorry, must have been leftover from an earlier iteration; removing

liliankasem commented on src/WebJobs.Script.Grpc/Server/FunctionRpcService.cs:
```diff
@@ -33,64 +31,48 @@ public FunctionRpcService(IScriptEventManager eventManager, ILogger<FunctionRpcS
+                }
 
-                if (await messageAvailable())
+                if (await await MoveNextAsync(requestStream, cancelSource))
```
I don't think I've used a double await before, would love to learn more about this pattern and why we do this

NickCraver replied:
The reason there's a double `await` here is the `Task<Task<T>>` nature, so we're awaiting the async yield of the task itself, and then (outer) awaiting _that _ task.

brettsam commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
-                .Timeout(_workerConfig.CountOptions.ProcessStartupTimeout)
-                .Take(1)
-                .Subscribe(SendWorkerInitRequest, HandleWorkerStartStreamError);
+            OnNext(MsgType.StartStream, _workerConfig.CountOptions.ProcessStartupTimeout, 1, SendWorkerInitRequest, HandleWorkerStartStreamError);
```
Is this `OnNext()` method here to line up with the Rx subscriptions we had set up earlier? A more lightweight subscription?

liliankasem replied:
I'm wondering if we can rename this to some thing more descriptive so its easier to follow, or the method that's using it

brettsam reviewed (commented):
I only have very general questions here. I think we should get this run on a private stamp to make sure we have no cold start regressions. 

I also want to take one more pass through to think through things like restarts and failures -- this is tricky!

liliankasem commented:
@mgravell anything we can do to help you get this PR merged?

fabiocav commented:
@mgravell indeed, this is high priority on our side, so whatever we can do to help, please let us know.

NickCraver commented:
I looked at this earlier on a call - it looks like the main pending item is to test regressions in cold start and not code tidy here as the blocker, or am I missing a thread where that's already in play? Happy to help on the code side.

liliankasem commented:
> I looked at this earlier on a call - it looks like the main pending item is to test regressions in cold start and not code tidy here as the blocker, or am I missing a thread where that's already in play? Happy to help on the code side.

I think it's a mix of both, would be great to get PR feedback addressed but not a huge blocker. @safihamid and @pragnagopa can you help with cold start testing?

safihamid commented:
> > I looked at this earlier on a call - it looks like the main pending item is to test regressions in cold start and not code tidy here as the blocker, or am I missing a thread where that's already in play? Happy to help on the code side.
> 
> I think it's a mix of both, would be great to get PR feedback addressed but not a huge blocker. @safihamid and @pragnagopa can you help with cold start testing?

Ok I will get cold start comparison numbers for this by end of next week. I hope most PR feedbacks are addressed by then.

mgravell committed:
Update src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs

Co-authored-by: Lilian Kasem <likasem@microsoft.com>

mgravell committed:
Update src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs

Co-authored-by: Lilian Kasem <likasem@microsoft.com>

mgravell committed:
nits and renames from PR review feedback

mgravell committed:
Merge branch 'mgravell/remove-rx' of https://github.com/Azure/azure-functions-host into mgravell/remove-rx

mgravell committed:
Merge branch 'dev' into mgravell/remove-rx

INCOMPLETE - still 2 new failing tests

# Conflicts:
#	src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs
#	test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs

mgravell commented:
Merge is in progress; fighting `SendInvocationRequest_ValidateTraceContext` and `SendInvocationRequest_ValidateTraceContext_SessionId` currently

safihamid commented:
No noticeable cold start changes with this PR in my private stamp as long as we are using the right language worker during placeholder mode. i.e. use Node placeholder for Node app, ...

mgravell committed:
Merge branch 'dev' into mgravell/remove-rx

# Conflicts:
#	src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs
#	test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs

mgravell committed:
fix tests on script-invocation context settings - was mock; now we simply check what actually arrives at the server

paulbatum commented:
I ran a few throughput tests in production using an App Service Plan running custom site extensions with and without the changes in this PR. 

TL;DR - while the gains for single core (P1V2) are marginal, the gains for multicore setups (P3V2 is 4-core) are significant, can be as high as 50% 🚀 🥇  

| Configuration | Throughput (dev) | Throughput (remove-rx) |
|-|-|-|
| P1V2, 100 connections | 401 RPS | 416 RPS |
| P3V2, 400 connections | 1042 RPS | 1581 RPS |

These tests were run with the following configuration:
- 30 second test runs
- workload is NET isolated anonymous HTTP triggered function that just returns a 200 response
- host.json is modified to reduce the amount of logging
- load is generated using bombadier e.g. 
`d:\tools\bombardier-windows-amd64.exe --insecure -c 400 -r 2400 -d 30s -m POST --format=pt -p i,r -l https://hostname.azurewebsites.net/api/function1`

dev branch latency distribution - 
```
Bombarding https://pbatum-rx-test-isolated.azurewebsites.net:443/api/function1 for 30s using 400 connection(s)
Statistics        Avg      Stdev        Max
  Reqs/sec      1042.60     200.06    1799.95
  Latency      380.22ms    36.21ms      1.28s
  Latency Distribution
     50%   376.40ms
     75%   388.20ms
     90%   408.00ms
     95%   429.02ms
     99%   506.01ms
  HTTP codes:
    1xx - 0, 2xx - 31666, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:   585.32KB/s
```

no-rx latency distribution - 
```
Bombarding https://pbatum-rx-test-isolated.azurewebsites.net:443/api/function1 for 30s using 400 connection(s)
Statistics        Avg      Stdev        Max
  Reqs/sec      1581.55     940.41    8996.40
  Latency      248.55ms    53.08ms      0.85s
  Latency Distribution
     50%   236.14ms
     75%   277.61ms
     90%   325.00ms
     95%   365.00ms
     99%   474.09ms
  HTTP codes:
    1xx - 0, 2xx - 47739, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:   845.26KB/s
```

NickCraver commented:
/azp run

NickCraver committed:
Merge branch 'dev' into mgravell/remove-rx

NickCraver committed:
Fix latest dev tests

NickCraver committed:
Fix new capabilities test to use channel properly

NickCraver commented:
Most failures here are due to more merges since the PR - 1 left I couldn't get through tonight:
```
[xUnit.net 00:03:48.32]     Microsoft.Azure.WebJobs.Script.Tests.WorkerConcurrencyManagerEndToEndTests.WorkerStatus_NewWorkerAdded [FAIL]
  Failed Microsoft.Azure.WebJobs.Script.Tests.WorkerConcurrencyManagerEndToEndTests.WorkerStatus_NewWorkerAdded [2 m]
  Error Message:
   System.ApplicationException : Condition not reached within timeout.
  Stack Trace:
     at Microsoft.Azure.WebJobs.Script.Tests.TestHelpers.Await(Func`1 condition, Int32 timeout, Int32 pollingInterval, Boolean throwWhenDebugging, Func`1 userMessageCallback) in /_/test/WebJobs.Script.Tests.Shared/TestHelpers.cs:line 115
   at Microsoft.Azure.WebJobs.Script.Tests.WorkerConcurrencyManagerEndToEndTests.WorkerStatus_NewWorkerAdded() in /_/test/WebJobs.Script.Tests.Integration/Rpc/WorkerConcurrencyManagerEndToEndTests.cs:line 28
--- End of stack trace from previous location ---
fail: Host.General[515]
      A host error has occurred during startup operation 'fe3a9a1d-e32b-4243-9d9f-2e2d3bcbbfe3'.
      System.ArgumentNullException: Value cannot be null. (Parameter 'provider')
         at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
         at Microsoft.Azure.WebJobs.Script.ScriptHostBuilderExtensions.<>c__DisplayClass6_0.<AddScriptHost>b__2(HostBuilderContext context, IConfigurationBuilder configBuilder) in /_/src/WebJobs.Script/ScriptHostBuilderExtensions.cs:line 114
         at Microsoft.Extensions.Hosting.HostBuilder.BuildAppConfiguration()
         at Microsoft.Extensions.Hosting.HostBuilder.Build()
         at Microsoft.Azure.WebJobs.Script.WebHost.DefaultScriptHostBuilder.BuildHost(Boolean skipHostStartup, Boolean skipHostConfigurationParsing) in /_/src/WebJobs.Script.WebHost/DefaultScriptHostBuilder.cs:line 59
         at Microsoft.Azure.WebJobs.Script.WebHost.WebJobsScriptHostService.BuildHost(Boolean skipHostStartup, Boolean skipHostJsonConfiguration) in /_/src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs:line 581
         at Microsoft.Azure.WebJobs.Script.WebHost.WebJobsScriptHostService.UnsynchronizedStartHostAsync(ScriptHostStartupOperation activeOperation, Int32 attemptCount, JobHostStartupMode startupMode) in /_/src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs:line 276
```

I'm not setup for end-to-end on laptop so can't debug readily but something is amiss there with services - will poke more tomorrow.

NickCraver committed:
Tidy up/PR comments

NickCraver committed:
Fix WorkerStatus_NewWorkerAdded by properly simulating latency in the new system

Here we're simulating latency in the outbound pipe was it was intended to when everything was posting through the script manager through RX. Also makes the latency non-static and more test friendly.

liliankasem commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -140,17 +134,133 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
+            }
+            catch
+            {
+                // TODO: log?
```
Yeah, let's add a log here so we can trace this if we get CRIs

NickCraver replied:
@liliankasem Can you please add the logging intended here? I'm not sure what the expectations are, probably quicker to push change to branch directly. It'll happen on _every_ gRPC event so not sure how we best handle potentially large volume things.

fabiocav replied:
Let's not block. @liliankasem maybe we can log an issue to track and have a follow up PR with anything that might be needed.

liliankasem replied:
Done! #8604

liliankasem reviewed (commented):
just noticed there's a TODO here, should probably resolve before merge! (+ any other TODOs I may have missed). If not addressed, you should create an issue for any items we're not handling in this PR

fabiocav commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -859,9 +958,53 @@ internal void HandleWorkerMetadataRequestError(Exception exc)
+                {
+                    try
+                    {
+                        pending.GetAwaiter().GetResult(); // ensure observed, for IVTS reasons
```
nit: would it be helpful to update the comment to something more descriptive (e.g., `// ensure observed to ensure IValueTaskSource completed/result is consumed`)?

NickCraver replied:
Updates!

fabiocav commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -102,31 +105,22 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
+                _outbound = outbound.Writer;
+                _inbound = inbound.Reader;
+                // note: we don't start the read loop until StartWorkerProcessAsync is called
+            }
```
The factory should create the channels prior to creating the `GrpcWorkerChannel`, so we should always have these channels setup here. Are expecting this to ever return `false`? Should we throw that is ever the case?

fabiocav commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -859,9 +958,53 @@ internal void HandleWorkerMetadataRequestError(Exception exc)
+        {
+            if (_outbound is null)
+            {
+                return default;
```
For this scenario, particularly if the message was an invocation request, shouldn't we fail here? In the case of an invocation, will will the behavior be if we simply return? I'm assuming the invocation would remain in-flight until the function timeout (which could be quite large) occurs.

NickCraver replied:
With a change above, we shouldn't be able to get in this state at all - going to remove these extra checks since it's no longer nullable and checked in the constructor.

NickCraver committed:
Merge remote-tracking branch 'origin/dev' into mgravell/remove-rx

NickCraver committed:
Address PR comments

Unsure if this covers #8604 completely or not - hope so!

NickCraver committed:
Test fix for SendInvocationRequest_SignalCancellation_WithCapability_SendsInvocationCancelRequest

NickCraver committed:
Fix PR comments - ensure _outbound and _inbound are never null

NickCraver committed:
Remove cruft

NickCraver commented:
Gentle reminder we should try and get this in before more conflicts pile up against the old script manager approach.

brettsam commented:
Let's get it in -- no objections from me. Although I do see conflicts again...

NickCraver committed:
Merge remote-tracking branch 'origin/dev' into mgravell/remove-rx

NickCraver committed:
Capture ObjectDisposedException in disposal race (occasional issue in tests)

NickCraver commented:
@brettsam merged once more from dev and addressed one test race (existing issue but hey, improvements!)

fabiocav commented:
Thanks, @NickCraver.

Good to have this merged once conflicts are resolved.

NickCraver commented:
@fabiocav I can't see merge conflicts local or PR UX, any idea what it's conflicting with?

fabiocav commented:
@NickCraver odd, I don't see the usual list, just the conflict warning in GH.

If you can merge locally, go for it!

NickCraver commented:
I beat the GitHub UI!
